hypervisor/dm/vpci/pci_pt.c

```diff
@@ -128,85 +132,56 @@ static void vdev_pt_remap_msix_table_vbar(struct pci_vdev *vdev)
+			uint64_t addr_hi, addr_lo;
+			struct mem_io_node *mmio_node = NULL;
+			struct acrn_vm *vm = vdev->vpci->vm;
+
 			/*
-			* PCI Spec: a BAR may also map other usable address space that is not associated
-			* with MSI-X structures, but it must not share any naturally aligned 4 KB
-			* address range with one where either MSI-X structure resides.
-			* The MSI-X Table and MSI-X PBA are permitted to co-reside within a naturally
-			* aligned 4 KB address range.
-			*
-			* If PBA or others reside in the same BAR with MSI-X Table, devicemodel could
-			* emulate them and maps these memory range at the 4KB boundary. Here, we should
-			* make sure only intercept the minimum number of 4K pages needed for MSI-X table.
-			*/
-
-			/* The higher boundary of the 4KB aligned address range for MSI-X table */
-			addr_hi = msix->mmio_gpa + msix->table_offset + (msix->table_count * MSIX_TABLE_ENTRY_SIZE);
-			addr_hi = round_page_up(addr_hi);
-
-			/* The lower boundary of the 4KB aligned address range for MSI-X table */
-			addr_lo = round_page_down(msix->mmio_gpa + msix->table_offset);
-		}
+			 * PCI Spec: a BAR may also map other usable address space that is not associated
+			 * with MSI-X structures, but it must not share any naturally aligned 4 KB
+			 * address range with one where either MSI-X structure resides.
+			 * The MSI-X Table and MSI-X PBA are permitted to co-reside within a naturally
+			 * aligned 4 KB address range.
+			 *
+			 * If PBA or others reside in the same BAR with MSI-X Table, devicemodel could
+			 * emulate them and maps these memory range at the 4KB boundary. Here, we should
+			 * make sure only intercept the minimum number of 4K pages needed for MSI-X table.
+			 */
+			if (old_mmio_gpa != 0UL) {
+				addr_lo = old_mmio_gpa + msix->table_offset;
+				addr_hi = addr_lo + (msix->table_count * MSIX_TABLE_ENTRY_SIZE);
+
+				/* The lower boundary of the 4KB aligned address range for MSI-X table */
+				addr_lo = round_page_down(addr_lo);
+				/* The higher boundary of the 4KB aligned address range for MSI-X table */
+				addr_hi = round_page_up(addr_hi);
+
+				mmio_node = unregister_mmio_emulation_handler(vm, addr_lo, addr_hi);
+				msix->mmio_gpa = 0UL;
+			}
 
-		if (vdev->bar_base_mapped[msix->table_bar] != addr_lo) {
-			(void)register_mmio_emulation_handler(vdev->vpci->vm, NULL,
-					vmsix_table_mmio_access_handler, addr_lo, addr_hi, vdev);
-			/* Remember the previously registered MMIO vbar base */
-			vdev->bar_base_mapped[msix->table_bar] = addr_lo;
+			if (mmio_gpa != 0UL) {
+				addr_lo = mmio_gpa + msix->table_offset;
+				addr_hi = addr_lo + (msix->table_count * MSIX_TABLE_ENTRY_SIZE);
+
+				/* The lower boundary of the 4KB aligned address range for MSI-X table */
+				addr_lo = round_page_down(addr_lo);
+				/* The higher boundary of the 4KB aligned address range for MSI-X table */
+				addr_hi = round_page_up(addr_hi);
+
+				if (register_mmio_emulation_handler(vm, mmio_node,
```

Comment:
mmio_node may be not initialized...

Reply:
@dongyaozu 
No. If it's the first register, the mmio_node will initialized to NULL which means we will get an availavle mmio_node form the mmio_node array. Otherwise, the mmio_node will pointer to its
unregistered mmio_node